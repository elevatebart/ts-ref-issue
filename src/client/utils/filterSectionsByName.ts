import getFilterRegExp from './getFilterRegExp';
import filterComponentsByName from './filterComponentsByName';
import * as Rsg from '../../typings';

/**
 * Fuzzy filters sections by section or component name.
 *
 * @param {Array} sections
 * @param {string} query
 * @return {Array}
 */
export default function filterSectionsByName(
	sections: Rsg.TOCItem[],
	query: string
): Rsg.TOCItem[] {
	const regExp = getFilterRegExp(query);

	return sections
		.map(section => {
			return {
				...section,
				sections: section.sections ? filterSectionsByName(section.sections, query) : [],
				components: section.components ? filterComponentsByName(section.components, query) : [],
			};
		})
		.filter(
			section =>
				section.components.length > 0 ||
				section.sections.length > 0 ||
				regExp.test(section.name || '-')
		);
}
