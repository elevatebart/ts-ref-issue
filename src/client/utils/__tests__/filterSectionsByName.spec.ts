import deepfreeze from 'deepfreeze';
import filterSectionsByName from '../filterSectionsByName';

const sections = deepfreeze([
	{
		name: 'General',
		sections: [
			{
				name: 'Particles',
				components: [
					{
						name: 'Button',
					},
					{
						name: 'Image',
					},
				],
			},
		],
	},
]);

describe('filterSectionsByName', () => {
	it('should recursively filter sections and components by name', () => {
		const result = filterSectionsByName(sections, 'button');
		expect(result).toMatchSnapshot();
	});

	it('should skip sections without matches inside', () => {
		const result = filterSectionsByName(sections, 'general');
		expect(result).toMatchSnapshot();
	});

	it('should return empty array if no components of sections match query', () => {
		const result = filterSectionsByName(sections, 'pizza');
		expect(result).toEqual([]);
	});
});
