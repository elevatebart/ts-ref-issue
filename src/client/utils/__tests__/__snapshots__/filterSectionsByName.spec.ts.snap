// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterSectionsByName should recursively filter sections and components by name 1`] = `
Array [
  Object {
    "components": Array [],
    "name": "General",
    "sections": Array [
      Object {
        "components": Array [
          Object {
            "name": "Button",
          },
        ],
        "name": "Particles",
        "sections": Array [],
      },
    ],
  },
]
`;

exports[`filterSectionsByName should skip sections without matches inside 1`] = `
Array [
  Object {
    "components": Array [],
    "name": "General",
    "sections": Array [],
  },
]
`;
